File: symfony/src/Service/Marketplace/TradeBidService.php
<?php

namespace App\Service\Marketplace;

use App\Entity\Marketplace\TradeBid;
use App\Repository\Marketplace\TradeBidRepository;

class TradeBidService
{
    public function __construct(
        private TradeBidRepository $repository,
    ) {}

    public function create(array $data): TradeBid
    {
        throw new \LogicException('Not implemented');
    }

    public function update(TradeBid $entity, array $data): TradeBid
    {
        throw new \LogicException('Not implemented');
    }

    public function outbidBy(int $id, $newAmount): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TradeBid not found: ' . $id);
        $result = $entity->outbidBy($newAmount);
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function retract(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TradeBid not found: ' . $id);
        $entity->retract();
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/marketplace/trade_bid_service.rb
module Marketplace
  class TradeBidService
    def initialize(repository = TradeBid)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def outbid_by(id)
      instance = TradeBid.find(id)
      result = instance.outbid_by(new_amount)
      instance.save!
      result
    end

    def retract(id)
      instance = TradeBid.find(id)
      instance.retract()
      instance.save!
    end
  end
end
